### PR TITLE
Remove assumption of reliable writer behavior in rtps_udp

### PR DESCRIPTION
**Pull request by mitza-oci**

### CODE REVIEW

File: dds/DCPS/RTPS/Spdp.cpp [hunk 1]
```diff
@@ -3114,11 +3120,21 @@ Spdp::associated() const
...
+Spdp::has_discovered_participant(const DCPS::RepoId& guid) const
 {
   return participants_.find(guid) != participants_.end();
 }
 
+ACE_CDR::ULong Spdp::get_participant_flags(const DCPS::RepoId& guid) const
+{
+  const DiscoveredParticipantMap::const_iterator iter = participants_.find(guid);
+  if (iter == participants_.end()) {
+    return 0;
```
Comment (jrw972) [inline dds/DCPS/RTPS/Spdp.cpp:3132]:
```suggestion
    return PFLAGS_EMPTY;
```

Comment (jrw972) [reply dds/DCPS/RTPS/Spdp.cpp:3132]: And so on for other cases of default initialization.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 2]
```diff
@@ -1758,8 +1759,7 @@ RtpsUdpDataLink::RtpsReader::process_heartbeat_i(const RTPS::HeartBeatSubmessage
 
   // Only valid heartbeats (see spec) will be "fully" applied to writer info
   if (!(hb_first < 1 || hb_last < 0 || hb_last < hb_first.previous())) {
-    // TODO: This check breaks interoperability.
-    if (writer->first_valid_hb_ && directed && heartbeat.readerId == id_.entityId) {
+    if (writer->first_valid_hb_ && (directed || !writer->sends_directed_hb())) {
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:1762]:
```suggestion
    if (writer->first_valid_hb_ && ((directed && heartbeat.readerId == id_.entityId) || !writer->sends_directed_hb())) {
```

**Review by jrw972 — CHANGES_REQUESTED** [review]

**Review by iguessthislldo — APPROVED** [review]

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp (hunk 2 above)
Comment (mitza-oci) [reply dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:1762]: This was deliberate.  The only calls to `RtpsReader::process_heartbeat_i` result from undirected or directed-to-this-Reader submessages.  The `directed` parameter is also unnecessary but I didn't remove it in this PR.

**Pull request merged**